contracts/governance/BaseWeightedMultisig.sol

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { IBaseWeightedMultisig } from '../interfaces/IBaseWeightedMultisig.sol';
+import { ECDSA } from '../libs/ECDSA.sol';
+
+abstract contract BaseWeightedMultisig is IBaseWeightedMultisig {
+    // keccak256('WeightedMultisig.Storage');
+    bytes32 private constant BASE_WEIGHTED_STORAGE_LOCATION =
+        0xa233fbcae4dcfad00091a9d8ff9561f12b3db9ec7227470684b4617d40a38746;
+
+    struct WeightedMultisigStorage {
+        uint256 currentSignersEpoch;
+        mapping(uint256 => bytes32) hashForSignersEpoch;
+        mapping(bytes32 => uint256) signersEpochForHash;
+    }
+
+    uint256 internal constant OLD_SIGNERS_RETENTION = 16;
+
+    /**********************\
+    |* External Functions *|
+    \**********************/
+
+    /*
+     * @dev This function checks if the provided signers are sorted and contain no duplicates
+     * @param signers The signers to check
+     * @return True if the signers are sorted and contain no duplicates
+     */
+    function currentSignersEpoch() external view returns (uint256) {
+        return _baseWeightedStorage().currentSignersEpoch;
+    }
+
+    /*
+     * @dev This function checks if the provided signers are sorted and contain no duplicates
+     * @param signers The signers to check
+     * @return True if the signers are sorted and contain no duplicates
+     */
+    function hashForSignersEpoch(uint256 epoch) external view returns (bytes32) {
+        return _baseWeightedStorage().hashForSignersEpoch[epoch];
+    }
+
+    /*
+     * @dev This function checks if the provided signers are sorted and contain no duplicates
+     * @param signers The signers to check
+     * @return True if the signers are sorted and contain no duplicates
+     */
+    function signersEpochForHash(bytes32 hash) external view returns (uint256) {
+        return _baseWeightedStorage().signersEpochForHash[hash];
+    }
+
+    /*************************\
+    |* Integration Functions *|
+    \*************************/
+    /*
+     * @dev This function takes messageHash and proof data and reverts if proof is invalid
+     * @param messageHash The hash of the message that was signed
+     * @param weightedSigners The weighted signers data
+     * @param signatures The signatures data
+     * @return isLatestSigners True if provided signers are the current ones
+     */
+    function _validateProof(
+        bytes32 messageHash,
+        bytes calldata weightedSigners,
+        bytes[] calldata signatures
+    ) internal view returns (bool isLatestSigners) {
+        WeightedMultisigStorage storage $ = _baseWeightedStorage();
```

Comment:
Any reason why you named the variable `$`? I've never seen that convention before

Reply:
I was trying to follow the convention suggested in the standard:
https://eips.ethereum.org/EIPS/eip-7201

Reply:
I agree that looks like jQuery 🤣  Yeah `storage` sounds good.